### PR TITLE
kafka-future-whencomplete

### PR DESCRIPTION
**Pull request by cmccabe**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by steven-aerts** [comment]
Looks nice, I have put my changes on top of this pull request.

I also added some documentation for whenComlete and changed which exception is propagated if both the future and the action complete exceptionally to line up with `CompletableFuture`.

I think we are still missing a test but that should not hold back the KIP voting I guess.

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 1]
```diff
@@ -146,6 +146,41 @@ R await(long timeout, TimeUnit unit)
...
+        @Override
+        public void accept(T val, Throwable exception) {
+            try {
+                if (exception != null) {
+                    biConsumer.accept(null, exception);
+                } else {
+                    biConsumer.accept(val, null);
+                }
+            } catch (Throwable e) {
+                exception = e;
```
Comment (steven-aerts) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:167]: So in my updated PR I change this line to line up with CompletableFuture.

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 2]
```diff
@@ -146,6 +146,41 @@ R await(long timeout, TimeUnit unit)
...
+            } else {
+                future.complete(val);
+            }
+        }
+    }
+
+    @Override
+    public <T> KafkaFuture<T> whenComplete(final BiConsumer<? super T, ? super Throwable> biConsumer) {
+        final KafkaFutureImpl<T> future = new KafkaFutureImpl<T>();
+        addWaiter(new WhenCompleteBiConsumer(future, biConsumer));
```
Comment (steven-aerts) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:180]: Compiler still gives some warning here, as you are erasing here the java generic type.  Adding the `<>`  infects the rest of your code with the `? super T` you have to add at the correct places.

Comment (steven-aerts) [reply clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:180]: Fixed this in my latest update of #4033.

**Comment by hachikuji** [comment]
Closing this since the changes were incorporated into #4033.

**Pull request closed without merging**